### ISSUE TITLE
Lag when using alSourcePlay on ThinkPad X1 Tablet 2nd Gen.

### ISSUE THREAD

**Issue by Slion** [kcat/openal-soft#302]
I got AL-Soft building and working on a Windows Lenovo ThinkPad Tablet.
However my frame rate takes a serious hit as soon as I use `alSourcePlay`.
`alSourcePlay` is being called directly from the rendering loop.
I can see AL-Soft is creating its own thread, as one would expect, so the frame time hit is possibly caused by `alSourcePlay` waiting on some mutex or something. I also get some crackling.

I'm totally new to OpenAL so I probably just need to be taught how to use it.
For this first draft implementation I used a single source for each buffer and I'm possibly trying to play it multiple times simultaneously.

I guess I could try swapping the DLLs with the ones from Creative to see if that helps. Still I see no reason why I should not be able to get descent performance from AL-Soft.

**Comment by Slion** [comment]
Using Creative's implementation fixes the issue.
Maybe I should build AL-Soft using a different configure.
I was using the tip of master.

**Comment by Slion** [comment]
> I should be able to try the Android build soon. I wonder what I will get there.

Android build runs flawlessly at 60 FPS. I'm a happy customer :smiley:. Still I ought to find a solution for that Windows problem.

**Comment by Slion** [comment]
It looks like the issue is down to whichever backend is being used.
When first reporting that issue I had all four backends enabled so I'm not sure which one was used:
* DSOUND
* WASAPI
* WAVE
* WINMM

I then tested only with DSOUND and although it lag issue could have been better the sounds were more broken.
Then I tested only with WINMM and the sound were playing with no so much of a lag it seems.
I guess I'll stick with WINMM for now.

**Comment by kcat** [comment]
Can you provide a trace log of the app? Set the `ALSOFT_LOGLEVEL` environment variable to `3` when you run the app, and it'll print a bunch of information to `stderr` that can be useful for me to figure out the problem.

Also, what are the CPU specs, and the audio device configuration? Laptops and tablets tend to use power-saving modes for the CPU which is known to cause problems for certain time-sensitive tasks (like real-time audio processing).

**Comment by Slion** [comment]
It's a ThinkPad X1 Tablet 2nd Gen.
CPU: Intel® Core™ i7-7Y75

Since I tried the same all backend configuration on a desktop workstation and it flies. Therefore the issue is indeed linked to that tablet hardware.

**Comment by Slion** [comment]
Here are the logs you requested.
WINMM backend is usable but still has some crackling noise and occasional performance hit.
WASAPI backend, which is being used when compiling all possible Windows backends, is not usable. Too much crackling noise, slow down distortions and huge performance impact.

<details>
  <summary>WINMM</summary>

AL lib: (II) Initializing library v1.19.1-c238619a master
AL lib: (II) Supported backends: winmm, null
AL lib: (II) Loading config C:\Users\micro\AppData\Roaming\alsoft.ini...
AL lib: (II) Got binary: C:\Dev\VisualStudio\bin\Prj-vs16\bin\Debug, Prj.exe
AL lib: (II) Loading config C:\Dev\VisualStudio\bin\Prj-vs16\bin\Debug\alsoft.ini...
AL lib: (II) Key disable-cpu-exts not found
AL lib: (II) Detected max CPUID function: 0x16 (ext. 0x80000008)
AL lib: (II) Vendor ID: ""
AL lib: (II) Name: "Intel(R) Core(TM) i7-7Y75 CPU @ 1.30GHz"
AL lib: (II) Extensions: +SSE +SSE2 +SSE3 +SSE4.1
AL lib: (II) Key rt-prio not found
AL lib: (II) Key resampler not found
AL lib: (II) Key trap-al-error not found
AL lib: (II) Key trap-alc-error not found
AL lib: (II) Key reverb/boost not found
AL lib: (II) Key drivers not found
AL lib: (II) Initialized backend "winmm"
AL lib: (II) Added "winmm" for playback
AL lib: (II) Added "winmm" for capture
AL lib: (II) Key excludefx not found
AL lib: (II) Key default-reverb not found
AL lib: (II) Got device "OpenAL Soft on Speaker/HP (Realtek High Defini", ID 0
AL lib: (II) Key channels not found
AL lib: (II) Key sample-type not found
AL lib: (II) Key frequency not found
AL lib: (II) Key period_size not found
AL lib: (II) Key periods not found
AL lib: (II) Key sources not found
AL lib: (II) Key slots not found
AL lib: (II) Key sends not found
AL lib: (II) Key ambi-format not found
AL lib: (II) Created device 000002E13CB5B3D0, "OpenAL Soft on Speaker/HP (Realtek High Defini"
AL lib: (II) Key hrtf not found
AL lib: (II) Pre-reset: Stereo, Float, 44100hz, 882 / 2646 buffer
AL lib: (WW) SSE performs best with multiple of 4 update sizes (662)
AL lib: (II) Post-reset: Stereo, Float, 44100hz, 662 / 2648 buffer
AL lib: (II) Key stereo-mode not found
AL lib: (II) Key cf_level not found
AL lib: (II) Key stereo-encoding not found
AL lib: (II) Stereo rendering
AL lib: (II) Enabling first-order ambisonic decoder
AL lib: (II) Channel config, Main: 3, Real: 2
AL lib: (II) Allocating 5 channels, 20480 bytes
AL lib: (II) Max sources: 256 (255 + 1), effect slots: 64, sends: 2
AL lib: (II) Front stablizer disabled
AL lib: (II) Key dither not found
AL lib: (II) Key dither-depth not found
AL lib: (II) Dithering disabled
AL lib: (II) Key output-limiter not found
AL lib: (II) Output limiter enabled, 0.0000dB limit
AL lib: (II) Fixed device latency: 997732ns
AL lib: (II) Key volume-adjust not found
AL lib: (II) Created context 000002E13CCEE100
</details>

<details>
  <summary>All backends</summary>

AL lib: (II) Initializing library v1.19.1-c238619a master
AL lib: (II) Supported backends: wasapi, dsound, winmm, null, wave
AL lib: (II) Loading config C:\Users\micro\AppData\Roaming\alsoft.ini...
AL lib: (II) Got binary: C:\Dev\VisualStudio\bin\Prj-vs16\bin\Debug, Prj.exe
AL lib: (II) Loading config C:\Dev\VisualStudio\bin\Prj-vs16\bin\Debug\alsoft.ini...
AL lib: (II) Key disable-cpu-exts not found
AL lib: (II) Detected max CPUID function: 0x16 (ext. 0x80000008)
AL lib: (II) Vendor ID: ""
AL lib: (II) Name: "Intel(R) Core(TM) i7-7Y75 CPU @ 1.30GHz"
AL lib: (II) Extensions: +SSE +SSE2 +SSE3 +SSE4.1
AL lib: (II) Key rt-prio not found
AL lib: (II) Key resampler not found
AL lib: (II) Key trap-al-error not found
AL lib: (II) Key trap-alc-error not found
AL lib: (II) Key reverb/boost not found
AL lib: (II) Key drivers not found
AL lib: (II) Starting message thread
AL lib: (II) Message thread initialization complete
AL lib: (II) Starting message loop
AL lib: (II) Initialized backend "wasapi"
AL lib: (II) Added "wasapi" for playback
AL lib: (II) Added "wasapi" for capture
AL lib: (II) Key excludefx not found
AL lib: (II) Key default-reverb not found
AL lib: (II) Got message "Open Device" (0x0000, this=0000026A9D3F0630)
AL lib: (II) Key channels not found
AL lib: (II) Key sample-type not found
AL lib: (II) Key frequency not found
AL lib: (II) Key period_size not found
AL lib: (II) Key periods not found
AL lib: (II) Key sources not found
AL lib: (II) Key slots not found
AL lib: (II) Key sends not found
AL lib: (II) Key ambi-format not found
AL lib: (II) Created device 0000026A9D3F56B0, "OpenAL Soft on Headphones (Bose QC35 II Stereo)"
AL lib: (II) Key hrtf not found
AL lib: (II) Pre-reset: Stereo, Float, 44100hz, 882 / 2646 buffer
AL lib: (II) Got message "Reset Device" (0x0001, this=0000026A9D3F0630)
AL lib: (II) Requesting playback format:
    FormatTag      = 0xfffe
    Channels       = 2
    SamplesPerSec  = 44100
    AvgBytesPerSec = 352800
    BlockAlign     = 8
    BitsPerSample  = 32
    Size           = 22
    Samples        = 32
    ChannelMask    = 0x3
    SubFormat      = {00000003-0000-0010-8000-00aa00389b71}
AL lib: (WW) SSE performs best with multiple of 4 update sizes (882)
AL lib: (II) Post-reset: Stereo, Float, 44100hz, 882 / 2646 buffer
AL lib: (II) Key stereo-mode not found
AL lib: (II) Key hrtf-paths not found
AL lib: (II) Key hrtf_tables not found
AL lib: (II) Searching C:\Dev\VisualStudio\bin\Gik-vs16\Legacy\Defense\group\*.mhr
AL lib: (II) Searching C:\Users\micro\AppData\Roaming\openal\hrtf\*.mhr
AL lib: (II) Searching C:\ProgramData\openal\hrtf\*.mhr
AL lib: (II) Got new file "Built-In 44100hz"
AL lib: (II) Adding built-in entry "Built-In 44100hz"
AL lib: (II) Got new file "Built-In 48000hz"
AL lib: (II) Adding built-in entry "Built-In 48000hz"
AL lib: (II) Key default-hrtf not found
AL lib: (II) Loading Built-In 44100hz...
AL lib: (II) Detected data set format v2
AL lib: (II) Loaded HRTF support for format: Stereo 44100hz
AL lib: (II) Key hrtf-mode not found
AL lib: (II) Full HRTF rendering enabled, using "Built-In 44100hz"
AL lib: (II) Skipped delay: 1, max delay: 28, new FIR length: 92
AL lib: (II) Key decoder/nfc not found
AL lib: (II) Channel config, Main: 4, Real: 2
AL lib: (II) Allocating 6 channels, 24576 bytes
AL lib: (II) Max sources: 256 (255 + 1), effect slots: 64, sends: 2
AL lib: (II) Front stablizer disabled
AL lib: (II) Key dither not found
AL lib: (II) Key dither-depth not found
AL lib: (II) Dithering disabled
AL lib: (II) Key output-limiter not found
AL lib: (II) Output limiter enabled, 0.0000dB limit
AL lib: (II) Fixed device latency: 997732ns
AL lib: (II) Got message "Start Device" (0x0002, this=0000026A9D3F0630)
AL lib: (II) Key volume-adjust not found
AL lib: (II) Created context 0000026A9D4033C0

</details>

As it is I'm planning to use the Creative implementation on Windows to workaround those issues.

**Comment by kcat** [comment]
Odd that WinMM and WASAPI get different devices. Nut either way, I think I probably see the issue with WASAPI. The audio device is detected as headphones which is making OpenAL Soft enable HRTF rendering, but a 1.3GHz CPU is really slow. You can try enabling a simpler HRTF mode by creating/editing `C:\Users\micro\AppData\Roaming\alsoft.ini` and adding the lines
```
[general]
hrtf-mode = ambi1
```
Or you can disable HRTF outright, by adding
```
[general]
hrtf = false
```

**Comment by Slion** [comment]
Mmmh, I may have forgotten to mentioned I  used BT headphones.

**Comment by Slion** [comment]
> 1.3GHz CPU is really slow
Well that's relative. Compare that to a [venerable Pentium 4](https://www.cpubenchmark.net/compare/Intel-i7-7Y75-vs-Intel-Pentium-4-1.60GHz/2880vs1054) and it does very well indeed, I mean that thing still has a boost frequency of 3.6 GHz 😃 

I did a quick and rough check of the app CPU usage when running using Creative's OpenAL and ALSoft here is what I got:
* Creative: 10-15%
* ALSoft: 45-50%

Assuming the app is running on a single CPU core it means it's indeed CPU bound when using ALSoft. Disabling HRTF appears to be making things a lot better even though the frame rate still takes a serious hit.

It seems there is room for optimisation but I won't bother looking into that just now as Creative's OpenAL does the job just fine on PC. Moreover, so far I've had no performance issues with ALSoft on Android which is the platform where it really needed. Though at some point energy consumption could be a concern on Android if CPU usage is not optimal. That's a topic for another issue however.

@kcat Thank you so much for your support.
